azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+        )
+        # The request response takes a tuple of actions
+        # It will call state based provider connect which
+        # It will add subscribe to pending action queue
+        # it will execute connect and then execute subscribe
+        # TODO : Should I have a waiting callback or a connack on which something happens ?
+        self._request_response_provider.connect(
+            self.on_connect_completed, post_connect_actions=[subscribe_action]
+        )
+
+    def _send_register_request(self, event_data):
+        logger.info("Sending registration request from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_REGISTRATION + rid, " ", callback=self.on_publish_completed
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _query_operation_status(self, event_data):
+        logger.info("Querying operation status from polling machine")
+        self._set_query_timer()
+
+        rid = str(uuid.uuid4())
+        registration_result = event_data.args[0].args[0]
+
+        operation_id = registration_result.operation_id
+        publish_action = PublishAction(
+            constant.PUBLISH_TOPIC_QUERYING_I
+            + rid
+            + constant.PUBLISH_TOPIC_QUERYING_II
+            + operation_id,
+            " ",
+            callback=self.on_publish_completed,
+        )
+        self._operations[rid] = publish_action
+        self._request_response_provider.send_request(publish_action)
+
+    def _handle_response_received(self, registration_result):
+        self._query_timer.cancel()
+        self._query_timer = None
+        if self.on_registration_status_update:
+            self.on_registration_status_update(registration_result)
+
+        if registration_result.status == "assigning":
+            rid = registration_result.request_id
+            if rid in self._operations:
+                publish_action = self._operations[rid]
+                # If this request was generated by us for a query operation,
+                # then we are getting a response for a query operation and
+                # and we must clear the polling timer before beginning anything
+                if "operationId" in publish_action.publish_topic:
+                    self._polling_timer.cancel()
+                    self._polling_timer = None
+            self._trig_wait(registration_result)
+        elif registration_result.status == "assigned":
+            self._trig_complete(registration_result)
+        else:
+            self._trig_error(registration_result)
+
+    def clear_and_disconnect(self, event_data):
+        self._query_timer = None
+        self._polling_timer = None
+        unsubscribe_action = UnsubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, callback=self.on_unsubscribe_completed
+        )
```

Comment:
Is it really necessary to unsubscribe here?  I'm not sure if there's a benefit and I'm not sure if the broker inside the gateway would do anything with this anyway.